cluster.c

```diff
@@ -871,48 +870,62 @@ RRStatus ShardingInfoAddShardGroup(RedisRaftCtx *rr, ShardGroup *new_sg)
     }
 
     return RR_OK;
+
+fail:
+    ShardGroupFree(sg);
+    return RR_ERROR;
 }
 
 /* Parse a ShardGroup specification as passed directly to RAFT.SHARDGROUP ADD.
  * Shard group syntax is as follows:
  *
  *  [shardgroup id] [num_slots] [num_nodes] ([start slot] [end slot] [slot type])* ([node-uid node-addr:node-port])*
  *
- * If parsing errors are encountered, an error reply is generated on the supplied RedisModuleCtx,
- * and RR_ERROR is returned.
+ * If parsing errors are encountered, an error reply is generated on the supplied RedisModuleCtx and -1 is returned
+ * If it was processed successfully, the number of argv elements consumed is returned
```

Comment:
Something doesn't add up here.
I'd expect to have only input validation here, but not validation against applied slot mapping configuration, because by the time we apply this it could change (due to other uncommitted changes). Am I missing something here?

Reply:
I don't understand.   That's what happens.  we do input validation here (are the values passed in sane)

Its called in 2 places

1) ShardGroupsParse (plural)

this is used by REPLACE, in here it validates that the set is sane (i.e. is internally consistent) as it will replace the whole thing.

2) RR_SHARDGROUP_ADD  RaftReqs where its passed to

where it will eventually be handled by handleShardGroupAdd() which because this isn't a whole world set of data, is validated relative to the state at apply time.

Reply:
i.e. a REPLACE operation doesn't depend on the eventual state at apply time, its replacing the entire state.